### DISCUSSION TITLE
Best way to simulate offline mode while developing?

### DISCUSSION THREAD

**Discussion by cmacdonnacha** [mswjs/msw#704] [Q&A] [answered]
I've been using MSW to mock my test data while developing locally. It's been great so far but I was wondering how I can simulate "offline mode" to see how my app responds to failed requests. Is it possible to use the Network tab in Chrome dev tools to do so?

![image](https://user-images.githubusercontent.com/9777763/115413923-807e1600-a1ed-11eb-8b56-7abd8ddaccda.png)

**Comment by kettanaito** [comment, marked as answer]
Hey, @cmacdonnacha.

You can simulate an offline mode using the `res.networkError()` utility.

```js
rest.get('/user', (req, res, ctx) => {
  return res.networkError('Your error message')
})
```

> Read more about [res.networkError()](https://mswjs.io/docs/api/response/network-error).

This produces an analogous response error as when requesting a resource while being offline.

**Reply by cmacdonnacha** [reply]
Thanks @kettanaito, wouldn't this always return a network error? Is there a way to have 200 OK returned for most cases but then return this network error only when "offline" is selected via chrome dev tools? 

**Full handler:**

```
res.get('/cars', (req, res, ctx) => {
  const manufacturer = req.url.searchParams.get('make')
  const cars = mockCars.filter(car => car.make === manufacturer)
  
  if (cars.length === 0) {
    return res(ctx.status(404))
  }

  return res(ctx.status(200), ctx.json(cars))
})
```

**Reply by kettanaito** [reply]
You can always return a conditional response in your resolver based on `navigator.onLine`:

```js
rest.get('/cars', (req, res, ctx) => {
  if (!navigator.onLine) {
    return res.networkError('Error message')
  }

  return res(ctx.json({ ... }))
})
```

**Reply by cmacdonnacha** [reply]
Oh nice, didn't even know about `navigator`. Thanks!

**Reply by cmacdonnacha** [reply]
@kettanaito Do you think it would be a good idea to update [these docs](https://mswjs.io/docs/api/response/network-error) to include this code? This way it's clear to the reader how they can use both network errors and regular 200 responses within a handler 

Example:

```
rest.get('/books', (req, res, ctx) => {
    if (!navigator.onLine) {
    return res.networkError('You are offline')
  }

  return res(ctx.json([{ name: 'The Great Gatsby' }]));
})
```

**Reply by kettanaito** [reply]
I think this is already mentioned in the [conditional response](https://mswjs.io/docs/basics/response-resolver#conditional-response) section. The fact that the condition may be anything: a custom logic, or a reference whether the page is offline—those are implied. I don't think we need an explicit example for this, but if many people find this necessary I don't mind adding it in the future (still, probably as a recipe).

**Reply by cmacdonnacha** [reply]
Yea I'm good with that, this was likely more me not being thorough enough reading through all the docs. The docs are really well put together and easy to follow, only piece of feedback I'd have is that it would be great if there was some search functionality.

**Reply by kettanaito** [reply]
We are working on a redesigned website that is going to include a search. We know it's painful to find the information in the current docs, and we will improve the experience.

**Reply by timdeschryver** [reply]
I just wanted to add that if you use the same handler in tests that run on Node.JS, you will have to provide a mock for `navigator` . Because I think that Node.JS doesn't have a `navigator` instance, and that it only exists in a browser context.
